File: ios/HoloNet/Module/Forum/Data/ForumCategoryRepository.swift
//
//  ForumCategoryRepository.swift
//  SWTOR HoloNet
//
//  Created by Ivan Fabijanovic on 16/10/14.
//  Copyright (c) 2014 Ivan Fabijanović. All rights reserved.
//

import UIKit
import Alamofire
import RxSwift
import RxAlamofire
import HTMLReader

protocol ForumCategoryRepository {
    func categories(language: ForumLanguage) -> Observable<[ForumCategory]>
    func categories(parent: ForumCategory) -> Observable<[ForumCategory]>
}

class DefaultForumCategoryRepository: ForumRepositoryBase, ForumCategoryRepository {
    func categories(language: ForumLanguage) -> Observable<[ForumCategory]> {
        return self.categories(id: language.rawValue)
    }
    
    func categories(parent: ForumCategory) -> Observable<[ForumCategory]> {
        return self.categories(id: parent.id)
    }
}

extension DefaultForumCategoryRepository {
    fileprivate func categories(id: Int) -> Observable<[ForumCategory]> {
        return self.manager.rx
            .string(.get, self.url(id: id))
            .map {
                let items = self.parse(html: $0)
                if items.isEmpty && self.isMaintenanceResponse($0) {
                    throw ForumError.maintenance
                }
                return items
            }
    }
    
    private func url(id: Int) -> URL {
        let string = "\(self.settings.forumDisplayUrl)?\(self.settings.categoryQueryParam)=\(id)"
        let url = URL(string: string)
        assert(url != nil)
        return url!
    }
    
    private func parse(html: String) -> [ForumCategory] {
        var items = [ForumCategory]()
        
        let document = HTMLDocument(string: html)
        let categoryNodes = document.nodes(matchingSelector: ".forumCategory > .subForum")
        
        for node in categoryNodes {
            if let category = self.parseCategory(element: node) {
                items.append(category)
            }
        }
        
        return items
    }
    
    private func parseCategory(element: HTMLElement) -> ForumCategory? {
        // Id & Title
        let titleElement = element.firstNode(matchingSelector: ".resultTitle > a")
        let idString = self.parser.linkParameter(linkElement: titleElement, name: self.settings.categoryQueryParam)
        let id = idString != nil ? Int(idString!) : nil
        let title = titleElement?.textContent
        
        // Icon
        var iconUrl: String? = nil
        if let thumbElement = element.firstNode(matchingSelector: ".thumbBackground") {
            if let iconStyle = thumbElement["style"] {
                let start = iconStyle.range(of: "url(", options: .literal, range: nil, locale: nil)
                let end = iconStyle.range(of: ")", options: .literal, range: nil, locale: nil)
                iconUrl = iconStyle.substring(with: Range(uncheckedBounds: (lower: start!.upperBound, upper: end!.lowerBound)))
            }
        }
        
        // Description
        let description = element.firstNode(matchingSelector: ".resultText")?.textContent
        
        // Stats & Last post
        var stats: String? = nil
        var lastPost: String? = nil
        let subTextElements = element.nodes(matchingSelector: ".resultSubText")

        if subTextElements.count > 0 {
            stats = subTextElements[0].textContent
        }
        if subTextElements.count > 1 {
            lastPost = subTextElements[1].textContent
        }
        
        if id == nil { return nil }
        if title == nil { return nil }
        
        let finalTitle = title!.stripNewLinesAndTabs().trimSpaces().collapseMultipleSpaces()
        let finalDescription = description?.trimSpaces()
        let finalStats = stats?.stripNewLinesAndTabs().trimSpaces().collapseMultipleSpaces()
        let finalLastPost = lastPost?.stripNewLinesAndTabs().trimSpaces().collapseMultipleSpaces()
        
        var category = ForumCategory(id: id!, title: finalTitle)
        category.iconUrl = iconUrl
        category.desc = finalDescription
        category.stats = finalStats
        category.lastPost = finalLastPost
        
        return category
    }
}

File: android/app/src/main/java/com/ifabijanovic/holonet/forum/data/ForumCategoryRepository.kt
package com.ifabijanovic.holonet.forum.data

import com.ifabijanovic.holonet.app.model.Settings
import com.ifabijanovic.holonet.forum.language.ForumLanguage
import com.ifabijanovic.holonet.forum.model.ForumCategory
import com.ifabijanovic.holonet.forum.model.ForumMaintenanceException
import com.ifabijanovic.holonet.helper.StringHelper
import io.reactivex.Observable
import io.reactivex.schedulers.Schedulers
import org.jsoup.Jsoup
import org.jsoup.nodes.Element

/**
 * Created by feb on 20/03/2017.
 */
interface ForumCategoryRepository {
    fun categories(language: ForumLanguage): Observable<List<ForumCategory>>
    fun categories(language: ForumLanguage, parent: ForumCategory): Observable<List<ForumCategory>>
}

class DefaultForumCategoryRepository(parser: ForumParser, service: ForumService, settings: Settings): ForumRepositoryBase(parser, service, settings), ForumCategoryRepository {
    override fun categories(language: ForumLanguage): Observable<List<ForumCategory>> {
        val localizedSettings = this.localizedSettings(language)
        return this.categories(localizedSettings.pathPrefix, localizedSettings.rootCategoryId)
    }

    override fun categories(language: ForumLanguage, parent: ForumCategory): Observable<List<ForumCategory>> {
        val localizedSettings = this.localizedSettings(language)
        return this.categories(localizedSettings.pathPrefix, parent.id)
    }

    private fun categories(language: String, categoryId: Int): Observable<List<ForumCategory>> {
        return this.service
                .category(language, categoryId)
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.computation())
                .map { html ->
                    val items = this.parse(html)
                    if (items.isEmpty() && this.isMaintenance(html)) {
                        throw ForumMaintenanceException()
                    }
                    return@map items
                }
    }

    private fun parse(html: String): List<ForumCategory> {
        val items = mutableListOf<ForumCategory>()

        Jsoup.parse(html)
                .select(".forumCategory > .subForum")
                .mapNotNullTo(items) { this.parseCategory(it) }

        return items.toList()
    }

    private fun parseCategory(element: Element): ForumCategory? {
        // Id & Title
        val titleElement = element.select(".resultTitle > a")?.first()
        val idString = this.parser.linkParameter(titleElement, this.settings.categoryQueryParam)
        var id: Int?
        try {
            id = idString?.toInt()
        } catch (e: Exception) {
            id = null
        }
        val title = titleElement?.text()

        // Icon
        var iconUrl: String? = null
        val thumbElement = element.select(".thumbBackground")?.first()
        if (thumbElement != null) {
            val iconStyle = thumbElement.attr("style")
            if (iconStyle != null) {
                val startString = "url("
                val start = iconStyle.indexOf(startString, 0, true)
                val end = iconStyle.indexOf(")", 0, true)
                iconUrl = iconStyle.substring(start + startString.count(), end)
            }
        }

        // Description
        val description = element.select(".resultText")?.first()?.text()

        // Stats & Last post
        var stats: String? = null
        var lastPost: String? = null
        val subTextElements = element.select(".resultSubText")

        if (subTextElements.size > 0) {
            stats = subTextElements[0].text()
        }

        if (subTextElements.size > 1) {
            lastPost = subTextElements[1].text()
        }

        if (id == null) { return null }
        if (title == null) { return null }

        val finalTitle = StringHelper(title).stripNewLinesAndTabs().trimSpaces().collapseMultipleSpaces().value
        val finalDescription = if (description != null) StringHelper(description).trimSpaces().value else null
        val finalStats = if (stats != null) StringHelper(stats).stripNewLinesAndTabs().trimSpaces().collapseMultipleSpaces().value else null
        val finalLastPost = if (lastPost != null) StringHelper(lastPost).stripNewLinesAndTabs().trimSpaces().collapseMultipleSpaces().value else null

        return ForumCategory(id, iconUrl, finalTitle, finalDescription, finalStats, finalLastPost)
    }
}
